src/mtype.h

```diff
@@ -268,8 +271,15 @@ class Type : public RootObject
     virtual bool isscope();
     virtual bool isString();
     virtual bool isAssignable();
+
+    virtual TypeStruct* isTypeStruct() { return NULL; }
+    virtual TypeClass* isTypeClass() { return NULL; }
+    virtual bool isTypeAggregate() { return this->isTypeStruct() || this->isTypeClass(); }
+    virtual TypeDArray* isTypeDArray() { return NULL; }
```

Comment:
Not quite... The user may make mistakes such as

```C
if (e->op == TOKtemplate)
{
    NonTemplateExp* ntx = (NonTemplateExp*)e;
}
```

which is not caught by the compiler.

Making mistakes with my alternative

if (TemplateExp* te = e->isTemplateExp())
{
    ...
}
```

will be caught by the compiler provided that isTemplateExp is implemented correctly which of course has to be dealt with but that's only in one place.